geo/src/main/java/org/apache/carbondata/geo/GeoHashUtils.java

```diff
@@ -0,0 +1,411 @@
+   * @param rangeListB geoId range list of polygonB
+   * @return geoId range list of processed set
+   */
+  public static List<Long[]> processRangeList(List<Long[]> rangeListA, List<Long[]> rangeListB,
+      String opType) {
+    List<Long[]> processedRangeList;
+    GeoOperationType operationType = GeoOperationType.getEnum(opType);
+    if (operationType == null) {
+      throw new RuntimeException("Unsupported operation type " + opType);
+    }
+    switch (operationType) {
+      case OR:
+        processedRangeList = getPolygonUnion(rangeListA, rangeListB);
+        break;
+      case AND:
+        processedRangeList = getPolygonIntersection(rangeListA, rangeListB);
+        break;
+      default:
+        throw new RuntimeException("Unsupported operation type " + opType);
+    }
+    return processedRangeList;
+  }
+
+  /**
+   * Get two polygon's union
+   *
+   * @param rangeListA geoId range list of polygonA
+   * @param rangeListB geoId range list of polygonB
+   * @return geoId range list after union
+   */
+  private static List<Long[]> getPolygonUnion(List<Long[]> rangeListA, List<Long[]> rangeListB) {
+    if (Objects.isNull(rangeListA)) {
+      return rangeListB;
+    }
+    if (Objects.isNull(rangeListB)) {
+      return rangeListA;
+    }
+    int sizeFirst = rangeListA.size();
+    int sizeSecond = rangeListB.size();
+    if (sizeFirst > sizeSecond) {
+      rangeListA.addAll(sizeFirst, rangeListB);
+      return mergeList(rangeListA);
+    } else {
+      rangeListB.addAll(sizeSecond, rangeListA);
+      return mergeList(rangeListB);
+    }
+  }
+
+  private static List<Long[]> mergeList(List<Long[]> list) {
+    if (list.size() == 0) {
+      return list;
+    }
+    Collections.sort(list, new Comparator<Long[]>() {
+      @Override
+      public int compare(Long[] arr1, Long[] arr2) {
+        return Long.compare(arr1[0], arr2[0]);
+      }
+    });
+    Long[] min;
+    Long[] max;
+    for (int i = 0; i < list.size(); i++) {
+      min = list.get(i);
+      for (int j = i + 1; j < list.size(); j++) {
+        max = list.get(j);
+        if (min[1] + 1 >= max[0]) {
+          min[1] = Math.max(max[1], min[1]);
+          list.remove(j);
+          j--;
+        }
+      }
+    }
+    return list;
+  }
+
+  /**
+   * Get two polygon's intersection
+   *
+   * @param rangeListA geoId range list of polygonA
+   * @param rangeListB geoId range list of polygonB
+   * @return geoId range list after intersection
+   */
+  private static List<Long[]> getPolygonIntersection(List<Long[]> rangeListA,
+      List<Long[]> rangeListB) {
+    List<Long[]> intersectionList = new ArrayList<>();
+    if (Objects.isNull(rangeListA) || Objects.isNull(rangeListB)) {
+      return Collections.emptyList();
+    }
+    int endIndex1 = rangeListA.size();
+    int endIndex2 = rangeListB.size();
+    int startIndex1 = 0;
+    int startIndex2 = 0;
+
+    while (startIndex1 < endIndex1 && startIndex2 < endIndex2) {
```

Comment:
```suggestion
while (startIndex1 < endIndex1 && startIndex2 < endIndex2) {
  long start = Math.max(rangeListA.get(startIndex1)[0], rangeListB.get(startIndex2)[0]);
  long end = Math.min(rangeListA.get(startIndex1)[1], rangeListB.get(startIndex2)[1]);
  if (start <= end) {
    intersectionList.add(new Long[] {start, end});
  }
  if (rangeListA.get(startIndex1)[1] < rangeListB.get(startIndex2)[1]) {
    startIndex1++;
  } else {
    startIndex2++;
  }
}
```

Reply:
this part code is developed by Discovery. `if` block code is no intersection part, `else `block code is intersection part, they think this `if else` may be more faster in some case, how do you think about it ?

Reply:
I have optimized the code with your suggestion